### PR TITLE
New documentation system

### PR DESCRIPTION
**Pull request by mikee47**
See [docs/README.md](https://sming.readthedocs.io/en/latest/_inc/docs/README.html) for building instructions.

You can find a pre-built version based on my repo. at https://sming.readthedocs.io. It should update automatically with any changes to my `feature/sphinx-docs` branch.

Please read [Contributing / Documentation](https://sming.readthedocs.io/en/latest/contribute/documentation.html) for details (source is `docs/source/contribute/documentation.rst`).

See #1683 for discussion.

### CODE REVIEW

File: docs/Makefile [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+endif
+endef # ParseComponent
+
+$(foreach d,$(COMPONENT_DIRS),$(eval $(call ParseComponent,$(notdir $d),$d)))
+
+
+#
+SOURCE_FILES	:= $(patsubst $(SMINGDIR)/%,%,$(SOURCE_FILES))
+
+SOURCE_DIRS		:= $(patsubst %/,$(SRC_INCDIR)/%,$(sort $(dir $(SOURCE_FILES))))
```
Comment (slaff) [inline docs/Makefile]: With the current code I ended up with the following error:
`make[1]: *** No rule to make target 'source/_inc/', needed by 'source/_inc/CONTRIBUTING.md'.  Stop.`
changing the line above to:
`SOURCE_DIRS		:= $(addsuffix /, $(patsubst %/,$(SRC_INCDIR)/%,$(sort $(dir $(SOURCE_FILES)))))`
fixed the issue(s) for me.

File: docs/source/conf.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+html_theme = 'sphinx_rtd_theme'
+
+# Add any paths that contain custom static files (such as style sheets) here,
+# relative to this directory. They are copied after the builtin static files,
+# so a file named "default.css" will overwrite the builtin "default.css".
+html_static_path = ['_static']
+
+
+##
+subprocess.run(["make", "-C", "../../Sming", "submodules", "SMING_ARCH=Host"])
```
Comment (slaff) [inline docs/source/conf.py]: Please, replace `subprocess.run` with `subprocess.call`. This way the code will work with python 2.7 and version 3 that is lower than 3.5.

**Comment by slaff** [comment]
@mikee47 we can add the initial version of the documentation. It is not complete but the structure looks good. Did you plan to add anything more here?

**Comment by mikee47** [comment]
@slaff Now the patching update is merged I'll rebase, tidy up and push the latest changes through. They're in my `dev/sphinx-docs` branch and that version is available on [readthedocs](https://sming.readthedocs.io/en/dev-sphinx-docs/) if you want to have a browse.

Rough list of changes:

* README files may be either `README.rst` or `README.md`
* Every Component, Library and Sample page gets an automatic 'references' section at the bottom, plus a list of any environment variables it defines. The actual README is included, and is now simpler as it doesn't need to bother with any of that.
* Fixes to doxygen comments (gets rid of a load of warnings/errors)
* Add section on patching to the 'building' docs.

**Comment by mikee47** [comment]
@slaff I'm reasonably OK with how the build system is working, so this morning taking some time to read through from the top and fix anything I come across. I'll check them in later today, up to you whether to include them in the initial merge or leave it for another PR.

**Comment by slaff** [comment]
@mikee47 With the latest update the list of sample project is not showing any longer. The list of Arduino Libraries is not showing either. What am I doing wrong?

**Comment by mikee47** [comment]
@slaff You've done nothing! RTD doesn't always track changes very well so I [wiped](https://docs.readthedocs.io/en/stable/guides/wipe-environment.html) and rebuilt it manually.

**Comment by mikee47** [comment]
@slaff Once you've merged this I'll update the RTD project to point to the real repo. I've added you as a maintainer so you should be able to make changes https://readthedocs.org/projects/sming/.

**Comment by slaff** [comment]
https://sming.readthedocs.io/en/latest/api/index.html: Concerning the API - is there something planned to integrate the current API docs (the ones that doxygen generates).

**Comment by mikee47** [comment]
Yes, more work required there. There should be stuff under [Streams](https://sming.readthedocs.io/en/latest/api/streams.html) now (reverted a change made during later updates) so that's the sort of thing. It does need breaking up more logically though so is a project all in itself.

**Comment by mikee47** [comment]
I've put back the original doxygen API, but it's bound up with Sphinx so still get version control.
I've also enabled graphs to  assist with navigating and understanding class hierarchies and other relationships.

Whilst there are a number of extensions which could help, there is no easy way to generate a similar level of information within Sphinx. In fact, the two sets of documentation are complimentary and serve different purposes.

Note that any code docs we pull into Sphinx using the breathe extension won't be linked to the API section.

**Comment by slaff** [comment]
@mikee47 I am planning to merge this PR. Did you plan to add something more or I should merge it today?

**Comment by mikee47** [comment]
I'm all finished for now. There is one (very) minor bug that I'm aware of, which affects markdown documents containing links to other (local) markdown. It's been raised already:

https://github.com/miyakogi/m2r/issues/41#issue-456867868

This affects the links to _Release Notes_ and _Contributors_ here:
https://sming.readthedocs.io/en/latest/_inc/Sming/Libraries/IR/index.html

**Comment by mikee47** [comment]
I've updated RTD, should have docs shortly.

**Comment by mikee47** [comment]
Couple of issues to fix re. #1763:

* Didn't pick up changes, so did a manual wipe/rebuild. Will investigate.
* Missing title in README for MCP23008. Probably needs clarifying in documentation.

I'll do a PR to address these.

**Comment by mikee47** [comment]
>     * Didn't pick up changes,

Can't see any issue with the build system, so I'm guessing it's because the webhook isn't working yet. I guess when it is we'll need to move things into another RTD account?

**Comment by slaff** [comment]
@mikee47  Do you want to submit as PR the work that you have done here: https://github.com/mikee47/Sming/tree/update/library-docs ?

**Comment by mikee47** [comment]
Still incomplete. Has uncovered some other stuff needs sorting as well.

**Pull request merged**